src/ninja.cc

```diff
@@ -1087,6 +1096,124 @@ int NinjaMain::ToolRestat(const Options* options, int argc, char* argv[]) {
   return EXIT_SUCCESS;
 }
 
+struct Compdb_Targets {
+  enum class Action {
+    display_help_and_exit,
+    emit_commands
+  };
+
+  Action action;
+  EvaluateCommandMode eval_mode = ECM_NORMAL;
+
+  std::vector<StringPiece> targets;
+
+  static Compdb_Targets parse(int argc, char* argv[]) {
+    //
+    // grammar:
+    //     ninja -t compdb-targets [-hx] target [targets]
+    //
+    Compdb_Targets ret;
+
+    // getopt_long() expects argv[0] to contain the name of
+    // the tool, i.e. "compdb".
+    argc++;
+    argv--;
+
+    // Phase 1: parse options:
+    {
+      optind = 1;  // see `man 3 getopt` for documentation on optind
+      int opt;
+      while ((opt = getopt(argc, argv, "hx")) != -1) {
+        switch (opt) {
+        case 'x':
+          ret.eval_mode = ECM_EXPAND_RSPFILE;
+          break;
+        case 'h':
+        default:
+          ret.action = Compdb_Targets::Action::display_help_and_exit;
+          goto fin;
+        }
+      }
+    }
+
+    // Phase 2: parse operands:
+    {
+      int targets_begin = optind;
+      int const targets_end = argc;
+
+      if (targets_begin == targets_end) {
+        Fatal("compdb-targets expects the name of at least one target");
+      } else {  // parse rules & targets (if any)
+        ret.action = Compdb_Targets::Action::emit_commands;
+        for (int i = targets_begin; i < targets_end; ++i)
+          ret.targets.push_back(argv[i]);
+      }
+    }
+
+  fin:
+    return ret;
+  }
+};
+
+
+void printCompdb(std::string const &directory, std::vector<Edge *> const &edges,
+                 const EvaluateCommandMode eval_mode) {
+  putchar('[');
+
+  bool first = true;
+  for (Edge* edge : edges) {
+    if (edge->is_phony() || edge->inputs_.empty())
+      continue;
+    if (!first)
+      putchar(',');
+    printOneCompdbObject(directory, edge, eval_mode);
+    first = false;
+  }
+
+  puts("\n]");
+}
+
+int NinjaMain::ToolCompilationDatabaseForTargets(const Options* options,
+                                                 int argc, char* argv[]) {
+  auto compdb = Compdb_Targets::parse(argc, argv);
+
+  switch (compdb.action) {
+  case Compdb_Targets::Action::display_help_and_exit: {
+    printf(
+        "usage: ninja -t compdb [-hx] target [targets]\n"
+        "\n"
+        "options:\n"
+        "  -h     display this help messsage\n"
+        "  -x     expand @rspfile style response file invocations\n");
+    return 1;
+  }
+
+  case Compdb_Targets::Action::emit_commands: {
+    InEdge_Collector inEdge_collector;
+
+    for (auto target_arg : compdb.targets) {
```

Comment:
An `for (auto var : range)` statement usually leads to copying every item during the iteration, which can have surprising effects, and is thus best avoided unless the type of `var` is trivial to see in the context of the code. However this is not the case here.

I recommend using `for (StringPiece target_arg : compdb.targets)` here, since it shows that the copy is harmless. Another option, in general, is to use `for (const auto& var : range)` if the iteration doesn't mutate the range items.
 
Use `StringPiece target_arg` here, since the `auto` is hard to resolve without looking way higher in the source code.